File: apps/ios/Exponential/UI/Onboarding/OnboardingDevicesStep.swift
import ExpUI
import ExpCore
import SwiftUI

/// Step 4 of the first-run wizard (EXP-725): point the user at the two places
/// a coding session can actually run.
///
/// LAST on purpose — it is the one step whose work happens on ANOTHER machine,
/// so it must not stand between a new user and their first board. Both
/// sub-cards reuse the getting-started checklist's copy (`desktop*` /
/// `server*`), because they are the same two steps said once: download the
/// desktop app, or copy the daemon install one-liner for an always-on box.
///
/// Under the cards, the caller's OWN registered machines off the synced
/// `devices` shape — the same rows (and the same settings sheet) the Devices
/// tab renders, so a machine that signed in while the wizard was open shows up
/// here and the trailing button turns from "Skip for now" into "Continue".
struct OnboardingDevicesStep: View {
    let accountId: String
    /// The wizard's trailing button, rendered by the host below this view —
    /// the step only reports whether at least one own machine exists.
    let onDevicesChanged: (Bool) -> Void

    @Environment(AppDependencies.self) private var deps
    @Environment(\.openURL) private var openURL

    @State private var viewModel: AgentsViewModel?
    @State private var settingsTarget: DeviceSettingsTarget?
    /// The pasteboard is silent — the server card's pill says so for 2s.
    @State private var copiedInstall = false
    @State private var copyFlashTask: Task<Void, Never>?

    /// The sheet's target is the ID only: the sheet reads the LIVE row itself
    /// (EXP-490), so a captured value would only go stale under it.
    private struct DeviceSettingsTarget: Identifiable {
        let id: String
    }

    /// Own machines only. The wizard never lists a teammate's shared server:
    /// this step is about the user's OWN setup.
    private var myDevices: [SteerDevice] {
        viewModel?.devices?.filter(\.isMine) ?? []
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            desktopCard
            serverCard

            GlassSectionHeader(OnboardingCopy.devicesYours)

            if myDevices.isEmpty {
                Text(OnboardingCopy.devicesNone)
                    .font(.caption)
                    .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                    .fixedSize(horizontal: false, vertical: true)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 12)
                    .glassRow()
            } else {
                ForEach(myDevices) { device in
                    deviceRow(device)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("onboarding-devices-step")
        .onAppear {
            if viewModel == nil {
                viewModel = AgentsViewModel(
                    accountId: accountId, userId: deps.auth.userId, db: deps.db
                )
            }
            viewModel?.startObserving()
        }
        .onDisappear {
            viewModel?.stopObserving()
            copyFlashTask?.cancel()
        }
        .onChange(of: myDevices.isEmpty) { _, empty in
            onDevicesChanged(!empty)
        }
        .sheet(item: $settingsTarget) { target in
            if let viewModel {
                // No team list: sharing a machine with a team is a later
                // decision, and the wizard's user is in exactly one team.
                DeviceSettingsSheet(
                    viewModel: viewModel, deviceId: target.id, teams: []
                )
            }
        }
    }

    // MARK: - The two install cards

    private var desktopCard: some View {
        installCard(
            icon: AppIcons.uiDevice,
            title: GettingStartedCopy.desktopTitle,
            description: GettingStartedCopy.desktopDescription,
            actionLabel: GettingStartedCopy.desktopAction
        ) {
            openURL(AppConstants.desktopReleasesUrl)
        }
    }

    private var serverCard: some View {
        installCard(
            icon: AppIcons.uiServer,
            title: GettingStartedCopy.serverTitle,
            description: GettingStartedCopy.serverDescription,
            actionLabel: copiedInstall
                ? OnboardingCopy.inviteCopied
                : GettingStartedCopy.serverAction
        ) {
            copyInstallSnippet()
        }
    }

    private func installCard(
        icon: String,
        title: String,
        description: String,
        actionLabel: String,
        action: @escaping () -> Void
    ) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 10) {
                AppIcon(icon, size: AppIcon.Size.medium)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                Text(title)
                    .font(.subheadline.weight(.medium))
                    .foregroundStyle(.white)
                    .lineLimit(1)
                Spacer(minLength: 8)
            }

            Text(description)
                .font(.caption)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)

            GlassPill(actionLabel, size: .sm, mode: .action(action))
        }
        .padding(10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .glassCard()
    }

    // MARK: - Machine rows

    /// The Devices tab's row, minus the launcher and the overflow menu — there
    /// is nothing to code on yet, and the wizard's only affordance is opening
    /// the settings sheet to sign an agent in.
    private func deviceRow(_ device: SteerDevice) -> some View {
        Button {
            settingsTarget = DeviceSettingsTarget(id: device.deviceId)
        } label: {
            HStack(spacing: 12) {
                AppIcon(
                    device.isServer ? AppIcons.uiServer : AppIcons.uiDevice,
                    size: AppIcon.Size.medium
                )
                .foregroundStyle(.white.opacity(TextOpacity.secondary))

                VStack(alignment: .leading, spacing: 3) {
                    Text(deviceName(device))
                        .font(.subheadline.weight(.medium))
                        .foregroundStyle(.white)
                        .lineLimit(1)
                    statusLine(device)
                }

                Spacer(minLength: 0)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 12)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .glassRow()
        .opacity(device.needsAgentSignIn ? 0.6 : 1)
    }

    /// The Devices tab's live/offline caption, minus the update states the
    /// wizard can never produce (EXP-409: signed-out agents replace "Online"
    /// when nothing is runnable, and annotate it otherwise).
    @ViewBuilder
    private func statusLine(_ device: SteerDevice) -> some View {
        let signedOut = device.unauthedAgentIds.joined(separator: ", ")
        let signInNeeded = device.needsAgentSignIn
        HStack(spacing: 5) {
            if device.isOnline {
                Circle()
                    .fill(signInNeeded ? DesignTokens.Semantic.yellow : DesignTokens.Semantic.green)
                    .frame(width: 6, height: 6)
                if signInNeeded {
                    Text("\(signedOut) not signed in")
                } else {
                    Text("Online")
                    if !signedOut.isEmpty {
                        Text("· \(signedOut) not signed in")
                            .foregroundStyle(.white.opacity(TextOpacity.quaternary))
                            .lineLimit(1)
                    }
                }
            } else {
                Text("Offline")
            }
        }
        .font(.caption)
        .foregroundStyle(.white.opacity(TextOpacity.tertiary))
    }

    private func deviceName(_ device: SteerDevice) -> String {
        device.deviceLabel.isEmpty ? device.deviceId : device.deviceLabel
    }

    private func copyInstallSnippet() {
        guard ServerInstallSnippet.copy(accountId: accountId, auth: deps.auth) else { return }
        copiedInstall = true
        copyFlashTask?.cancel()
        copyFlashTask = Task {
            try? await Task.sleep(for: .seconds(2))
            if !Task.isCancelled { copiedInstall = false }
        }
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/onboarding/OnboardingDevicesStep.kt
package com.exponential.app.ui.onboarding

import android.content.Intent
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.exponential.app.AppConstants
import com.exponential.app.data.api.SteerDevice
import com.exponential.app.ui.components.GlassPill
import com.exponential.app.ui.components.GlassSubmitButton
import com.exponential.app.ui.gettingstarted.GettingStartedCopy
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.issue.NeedsInputAmber
import com.exponential.app.ui.issue.ReviewGreen
import com.exponential.app.ui.issue.StaticDot
import com.exponential.app.ui.issue.relativeTime
import com.exponential.app.ui.session.DeviceSettingsSheet
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.glassCard
import com.exponential.app.ui.theme.glassRow

/**
 * Wizard step 4 — "Set up your devices" (EXP-725), the LAST step because
 * finishing it means walking over to another machine.
 *
 * Two install cards (the desktop IDE, the headless CLI daemon) reusing the
 * getting-started checklist's words verbatim — one copy for the two surfaces
 * that say the same thing — then the caller's own machines as they arrive over
 * the synced devices shape. A row taps into the same [DeviceSettingsSheet] the
 * Agents tab opens, so signing an agent in never needs a detour.
 *
 * Skippable: the trailing button reads [OnboardingCopy.SKIP] until a machine
 * of the user's own has registered, then [OnboardingCopy.CONTINUE].
 */
@Composable
fun OnboardingDevicesStep(
    instanceOrigin: String?,
    onContinue: () -> Unit,
    viewModel: OnboardingDevicesViewModel = hiltViewModel(),
) {
    val devices by viewModel.devices.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val clipboard = LocalClipboardManager.current
    var settingsTarget by remember { mutableStateOf<String?>(null) }
    var copied by remember { mutableStateOf(false) }
    LaunchedEffect(copied) {
        if (copied) {
            kotlinx.coroutines.delay(2_000)
            copied = false
        }
    }

    Column(
        modifier = Modifier
            .widthIn(max = 460.dp)
            .fillMaxWidth()
            .testTag("onboarding-devices-step"),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            OnboardingCopy.DEVICES_TITLE,
            style = MaterialTheme.typography.headlineSmall,
            color = MaterialTheme.colorScheme.onSurface,
            textAlign = TextAlign.Center,
        )
        Spacer(Modifier.height(12.dp))
        Text(
            OnboardingCopy.DEVICES_SUBTITLE,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
            textAlign = TextAlign.Center,
        )
        Spacer(Modifier.height(28.dp))

        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            InstallCard(
                title = GettingStartedCopy.DESKTOP_TITLE,
                description = GettingStartedCopy.DESKTOP_DESCRIPTION,
                actionLabel = GettingStartedCopy.DESKTOP_ACTION,
                actionIcon = ExpIcons.uiDownload,
                onAction = {
                    runCatching {
                        context.startActivity(
                            Intent(Intent.ACTION_VIEW, AppConstants.DESKTOP_RELEASES_URL.toUri()),
                        )
                    }
                },
            )
            // No resolved origin means no instance to point the daemon at —
            // copying `EXP_INSTANCE= sh` would hand over a broken command
            // (the Issues-tab checklist gates the same way).
            InstallCard(
                title = GettingStartedCopy.SERVER_TITLE,
                description = GettingStartedCopy.SERVER_DESCRIPTION,
                actionLabel = if (copied) OnboardingCopy.INVITE_COPIED else GettingStartedCopy.SERVER_ACTION,
                actionIcon = if (copied) ExpIcons.uiCheck else ExpIcons.uiCopy,
                onAction = instanceOrigin?.let { origin ->
                    {
                        clipboard.setText(AnnotatedString(AppConstants.serverInstallSnippet(origin)))
                        copied = true
                    }
                },
            )

            Text(
                OnboardingCopy.DEVICES_YOURS,
                style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold),
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.fillMaxWidth(),
            )
            val rows = devices.orEmpty()
            if (rows.isEmpty()) {
                Text(
                    OnboardingCopy.DEVICES_NONE,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
                    modifier = Modifier
                        .fillMaxWidth()
                        .glassRow()
                        .padding(horizontal = 12.dp, vertical = 10.dp),
                )
            } else {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    rows.forEach { device ->
                        OwnDeviceRow(
                            device = device,
                            onClick = { settingsTarget = device.deviceId },
                        )
                    }
                }
            }
        }

        Spacer(Modifier.height(28.dp))
        // A machine of one's own is what this step is for; until one shows up
        // the only honest label is "skip".
        GlassSubmitButton(
            label = if (devices.orEmpty().isEmpty()) OnboardingCopy.SKIP else OnboardingCopy.CONTINUE,
            onClick = onContinue,
        )
    }

    // Same live re-resolution as the Agents tab: the row can vanish mid-edit
    // (removed elsewhere), and the sheet simply stops rendering.
    settingsTarget?.let { targetId ->
        devices?.firstOrNull { it.deviceId == targetId && it.isMine }?.let { target ->
            DeviceSettingsSheet(
                device = target,
                onDismiss = { settingsTarget = null },
            )
        }
    }
}

/** One install card: title, one line of why, one pill. */
@Composable
private fun InstallCard(
    title: String,
    description: String,
    actionLabel: String,
    actionIcon: androidx.compose.ui.graphics.vector.ImageVector,
    onAction: (() -> Unit)?,
) {
    Column(
        modifier = Modifier.fillMaxWidth().glassCard().padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text(
            title,
            style = MaterialTheme.typography.titleSmall.copy(fontWeight = FontWeight.SemiBold),
            color = MaterialTheme.colorScheme.onSurface,
        )
        Text(
            description,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
        )
        GlassPill(label = actionLabel, icon = actionIcon, onClick = onAction, enabled = onAction != null)
    }
}

/**
 * One of the caller's machines: kind glyph, label, and the Agents-tab presence
 * caption (green dot Online / amber "<agents> not signed in" / "Last seen …" /
 * Offline). Tapping opens the device settings sheet.
 */
@Composable
private fun OwnDeviceRow(device: SteerDevice, onClick: () -> Unit) {
    val online = device.online
    val unauthed = device.unauthedAgentIds
    val signInNeeded = online && !device.hasRunnableAgent && unauthed.isNotEmpty()
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .glassRow()
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            if (device.isServer) ExpIcons.uiServer else ExpIcons.uiDevice,
            contentDescription = null,
            modifier = Modifier.size(18.dp),
            tint = MaterialTheme.colorScheme.onSurface.copy(
                alpha = if (online && !signInNeeded) TextEmphasis.Secondary else TextEmphasis.Tertiary,
            ),
        )
        Spacer(Modifier.width(12.dp))
        Column(modifier = Modifier.weight(1f)) {
            Text(
                device.deviceLabel.ifBlank { device.deviceId },
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                if (online) StaticDot(if (signInNeeded) NeedsInputAmber else ReviewGreen, size = 6.dp)
                Text(
                    when {
                        signInNeeded -> "${unauthed.joinToString(", ")} not signed in"
                        online -> "Online"
                        device.lastSeenAt != null -> "Last seen ${relativeTime(device.lastSeenAt)}"
                        else -> "Offline"
                    },
                    style = MaterialTheme.typography.bodySmall,
                    color = if (signInNeeded) {
                        NeedsInputAmber
                    } else {
                        MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary)
                    },
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}
